cores/esp8266/WString.h

```diff
@@ -279,7 +278,7 @@ class String {
 
     protected:
         // Contains the string info when we're not in SSO mode
-        struct _ptr { 
+        struct _ptr {
             char *   buff;
```

Comment:
Just curious.
Why are those members not initialized to null_ptr and 0?

I guess if the allocation of a buffer fails, it could be anything in there and when the destructor is called it will try to destruct something that's not to be destructed -> crash.
As we have seen while using the std::list emplace function, this could happen.
My work around for now is to emplace an empty object and then fill it while being in the list already.

Reply:
@TD-er, I'm not following you here.

This is 1/2 of the union{} and neither this struct nor the _sso one have any C++-type new/delete operations associated with them.  Iff the string was not in SSO mode, this buff will be freed in the ~String().  The union is initted in the String() constructor.

Reply:
@earlephilhower Every now and then I'm still looking for the explanation why the String class caused crashes when being destructed like I showed you before.
And the only thing I can think of is that there must be some possible code path where sso is set to false but the buffer pointer is not set due to OOM allocation while the String object is constructed.
So far I have not found any possible code path that could lead to this, but apparently there must be one as it otherwise could not cause these crashes while being destructed on members of a larger struct created using the emplace function of a list.
But even though I cannot find the code path to it, it would make sense for the buffer pointer to always being set to `nullptr` for every occasion where sso is set.

Setting it at the construction of the `_prt` struct is one thing and I agree with you that it is no guarantee it will be correct as it shares the memory with the `_sso` struct.

So maybe this function could be extended to something like this:
```c++
inline void setSSO(bool set) { 
    if (sso.isSSO && !set) {
        ptr.buff = nullptr;
    }
    sso.isSSO = set;
}
```

But on the other hand the only places where `setSSO(false)` is called, also the `setBuffer` is set.
Of those very few calls, only the call to `changeBuffer` can be slightly suspicious.
In which a call to `realloc` is made which should return either `nullptr` or a valid pointer.


Maybe a bit long way of me to show I don't like the open end where I simply cannot explain the crashes I had and have been working around them since. But I am always afraid I will one day forget to do so or merge a PR which doesn't take it into account and thus run into those hard to reproduce issues again.
So still looking at each and every change made to the String class to hopefully one day see something that may explain those issues.

Reply:
This SSO optimization was moved into the ESP32 repo as well, and they don't seem to be reporting issues related to OOM cases.  Not to say there's no bugs, but if there are we don't seem to be hitting them in other folks' code.

`free()` is an interesting beast because UMM doesn't keep anything like a list of allocated blocks.  They're just a few bits in memory right before the returned `malloc()` value.  So if there has been memory overwrite of even a byte past the end of the previous pointer, really weird things can happen.  But that's something you'd need `valgrind` or similar to detect at runtime (and that's not gonna run on the 8266!).

If you think there should be a reset in String, drop a PR and we can see about it.  Actually, if you think that's the issue it might be worthwhile to change your own version and run your code to see if it fixes the crashes.  That would be a good smoking gun test case (although it would be best to find and repro a failing case in the host unit tests, of course).

Good luck!

Reply:
Just a thought.

The issues I saw were with (multiple) String objects as members of a larger struct.
What would happen if the OOM occurred when constructing this larger struct?
What would the state be of those String objects? Has the constructor run?
If the constructor has not yet run, the init() function is not called, but the members are there.
So would it be useful to at least intialize the `isSSO` member of the `_sso` struct?
Like this:
```c++
        struct _sso {
            char     buff[SSOSIZE] = { 0 };
            unsigned char len   : 7 = 0; // Ensure only one byte is allocated by GCC for the bitfields
            unsigned char isSSO : 1 = 1;
        } __attribute__((packed)); // Ensure that GCC doesn't expand the flag byte to a 32-bit word for alignment issues
```

This to make sure the desctructor is not doing funny things as it needs `isSSO` to be 0 to call `free` on the buffer.
```c++
String::~String() {
    if(!isSSO() && wbuffer())
        free(wbuffer());
}
```

Reply:
@TD-er:
> Like this:

+1
this also eliminates explicit `init()` from constructors

Reply:
See #7590